File: src/nodes/synthesizer.py
import os
import re
import json
from typing import Dict, Any
from langchain_core.messages import SystemMessage, HumanMessage
from src.state import GlobalState
from src.llm import get_llm
from src.prompts.synthesizer_prompt import THEME_EXTRACTOR_PROMPT, HTML_SHELL_TEMPLATE
from src.utils.token_tracker import log_token_usage

def extract_root_css(response_text: str) -> str:
    """
    Extracts the CSS block containing :root definition from the model's response.
    """
    css_match = re.search(r'```css\s*\n(.*?)```', response_text, re.DOTALL)
    if css_match:
        return css_match.group(1).strip()
    
    # Fallback to general code blocks if specific language tag is missing
    code_match = re.search(r'```\s*\n(.*?)```', response_text, re.DOTALL)
    if code_match:
        return code_match.group(1).strip()
        
    # If no blocks found, return cleaned original text
    return response_text.strip()

def run_synthesizer(state: GlobalState) -> Dict[str, Any]:
    """
    LangGraph node: Combines HTML and CSS from all coded sections into a single index.html.
    Extracts Google Fonts and generates design tokens via Gemini.
    Writes the final file to output/index.html.
    """
    coded_sections = state.get("coded_sections", [])
    enhanced_prompt = state.get("enhanced_prompt", "")
    user_prompt = state.get("user_prompt", "AI Generated UI Layout")
    
    if not coded_sections:
        print("\n[Synthesizer] No coded sections to synthesize!")
        return {
            "pipeline_status": "failed",
            "failure_reason": "No coded sections found to synthesize",
        }
        
    # Filter to only successfully coded sections
    successful_sections = [s for s in coded_sections if s["status"] == "success"]
    
    print("\n[Synthesizer] Generating global design tokens and theme...")
    
    # Initialize Gemini to derive theme tokens
    try:
        llm = get_llm(temperature=0.4)
        theme_messages = [
            SystemMessage(content=THEME_EXTRACTOR_PROMPT),
            HumanMessage(content=f"UI Request: {user_prompt}\n\nDetailed Context: {enhanced_prompt}")
        ]
        theme_response = llm.invoke(theme_messages)
        log_token_usage("Theme Extractor", theme_response)
        root_css = extract_root_css(theme_response.content)
    except Exception as e:
        print(f"[Synthesizer] ⚠ Theme generation failed ({e}). Using default theme tokens.")
        # Fallback theme tokens
        root_css = """:root {
  --bg-primary: #0f172a;
  --bg-secondary: #1e293b;
  --bg-tertiary: #334155;
  --text-primary: #f8fafc;
  --text-secondary: #94a3b8;
  --accent-color: #3b82f6;
  --accent-hover: #2563eb;
  --font-family: 'Inter', sans-serif;
  --font-heading: 'Inter', sans-serif;
  --spacing-xs: 4px;
  --spacing-sm: 8px;
  --spacing-md: 16px;
  --spacing-lg: 32px;
  --spacing-xl: 64px;
  --border-radius: 8px;
  --border-radius-lg: 16px;
}"""    # Assemble CSS blocks to get a combined CSS string for scanning fonts
    css_blocks = []
    for sec_code in successful_sections:
        css_blocks.append(f"/* Section {sec_code['section_id']}: {sec_code['section_name']} */\n{sec_code['css']}")

    # Extract Google Fonts to link/import from both root_css and section CSS blocks
    all_css_to_scan = root_css + "\n" + "\n".join(css_blocks)
    font_matches = re.findall(
        r"(?:font-family|font-heading|font-mono|--font-[a-z0-9-]+):\s*['\"]?([^'\"\;,\)]+)['\"]?",
        all_css_to_scan
    )
    
    google_fonts_import = ""
    if font_matches:
        font_families = []
        for font in set(font_matches):
            font = font.split(",")[0].strip()
            # Clean quotes if any
            font = font.strip("'\"")
            if font.lower() not in ["sans-serif", "serif", "monospace", "cursive", "system-ui", "-apple-system", "inherit", "initial", "unset"] and not font.startswith("var("):
                font_families.append(font.replace(" ", "+"))
        if font_families:
            font_query = "|".join(font_families)
            google_fonts_import = f"@import url('https://fonts.googleapis.com/css2?family={font_query}:wght@400;600;700;800&display=swap');\n\n"

    print("[Synthesizer] Assembling React components and stylesheets...")
    
    jsx_components = []
    resolved_cms = {}
    db_records = []
    
    for sec_code in successful_sections:
        section_id = sec_code["section_id"]
        section_name = sec_code["section_name"]
        
        # Get raw JSX, strip internal React imports to prevent duplicate or invalid imports
        raw_jsx = sec_code["jsx"]
        clean_jsx = re.sub(r'^\s*import\s+.*?;?\s*$', '', raw_jsx, flags=re.MULTILINE)
        clean_jsx = clean_jsx.replace("export default function", "function")
        jsx_components.append(f"// --- Section {section_id}: {section_name} ---\n{clean_jsx}")
        
        # CMS Resolution: find the section CMS from state.sections
        section_state = next((s for s in state.get("sections", []) if s["id"] == section_id), None)
        if section_state and section_state.get("cms"):
            cms = section_state["cms"]
            db_records.append(cms)
            
            # Format elements as key-value for easy JSX rendering
            # sectionName -> frequentlyAskedQuestions
            component_key = "".join(word.capitalize() for word in re.split(r'[^a-zA-Z0-9]', section_name))
            component_key = component_key[0].lower() + component_key[1:]
            
            elements_map = {}
            for elem in cms.get("elements", []):
                elem_name = elem.get("elementName")
                if elem_name:
                    elements_map[elem_name] = elem
                    
            resolved_cms[component_key] = elements_map

    # Build the combined JSX App Component
    wrapper_lines = [
        "export function GeneratedPage({ cmsData }) {",
        "  return (",
        "    <div className=\"generated-page-container\">"
    ]
    
    for sec_code in successful_sections:
        section_name = sec_code["section_name"]
        component_name = "".join(word.capitalize() for word in re.split(r'[^a-zA-Z0-9]', section_name))
        prop_key = component_name[0].lower() + component_name[1:]
        wrapper_lines.append(f"      <{component_name} cmsData={{cmsData.{prop_key}}} />")
        
    wrapper_lines.append("    </div>")
    wrapper_lines.append("  );")
    wrapper_lines.append("}")

    # Determine which React hooks are referenced across all generated JSX components
    all_jsx_content = "\n".join(jsx_components)
    hooks_to_import = ["useState", "useMemo"]
    possible_hooks = [
        "useEffect", "useRef", "useCallback", "useContext", "useReducer",
        "useImperativeHandle", "useLayoutEffect", "useDebugValue"
    ]
    for hook in possible_hooks:
        if re.search(r'\b' + hook + r'\b', all_jsx_content):
            hooks_to_import.append(hook)
    
    hooks_import_list = ", ".join(sorted(list(set(hooks_to_import))))
    
    combined_jsx = (
        f"import React, {{ {hooks_import_list} }} from 'react';\n"
        "import cmsDataRaw from './cms_data.json';\n"
        "import './index.css';\n\n"
        + "\n\n".join(jsx_components) + "\n\n"
        + "\n".join(wrapper_lines) + "\n\n"
        "export default function App() {\n"
        "  return <GeneratedPage cmsData={cmsDataRaw.resolved_cms} />;\n"
        "}"
    )
    combined_css = google_fonts_import + root_css + "\n\n" + "\n\n".join(css_blocks)
    
    # Save CMS payloads
    cms_payload = {
        "db_records": db_records,
        "resolved_cms": resolved_cms
    }
    
    project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    
    # Paths for output folder
    output_dir = os.path.join(project_root, "output")
    os.makedirs(output_dir, exist_ok=True)
    out_jsx_file = os.path.join(output_dir, "page.jsx")
    out_css_file = os.path.join(output_dir, "page.css")
    out_cms_file = os.path.join(output_dir, "cms_data.json")
    
    # Paths for testing_react project
    react_src_dir = os.path.join(project_root, "testing_react", "src")
    os.makedirs(react_src_dir, exist_ok=True)
    react_jsx_file = os.path.join(react_src_dir, "App.jsx")
    react_css_file = os.path.join(react_src_dir, "index.css")
    react_cms_file = os.path.join(react_src_dir, "cms_data.json")
    
    try:
        # Write to output folder
        with open(out_jsx_file, "w", encoding="utf-8") as f:
            f.write(combined_jsx)
        with open(out_css_file, "w", encoding="utf-8") as f:
            f.write(combined_css)
        with open(out_cms_file, "w", encoding="utf-8") as f:
            json.dump(cms_payload, f, indent=2)
            
        # Write directly to testing_react project
        with open(react_jsx_file, "w", encoding="utf-8") as f:
            f.write(combined_jsx)
        with open(react_css_file, "w", encoding="utf-8") as f:
            f.write(combined_css)
        with open(react_cms_file, "w", encoding="utf-8") as f:
            json.dump(cms_payload, f, indent=2)
            
        print(f"[Synthesizer] ✓ React components compiled to output: {out_jsx_file}")
        print(f"[Synthesizer] ✓ CSS stylesheet written to output: {out_css_file}")
        print(f"[Synthesizer] ✓ Unified CMS JSON payload written to output: {out_cms_file}")
        print(f"[Synthesizer] ✓ React components integrated into testing_react: {react_jsx_file}")
        print(f"[Synthesizer] ✓ CSS stylesheet integrated into testing_react: {react_css_file}")
        print(f"[Synthesizer] ✓ CMS JSON integrated into testing_react: {react_cms_file}")
    except Exception as e:
        print(f"[Synthesizer] ✗ Failed to write output files: {e}")
        return {
            "pipeline_status": "failed",
            "failure_reason": f"Writing synthesized files failed: {e}",
        }
        
    return {
        "final_html": combined_jsx,  # Kept in state for graph output
        "output_path": out_jsx_file,
        "pipeline_status": "complete",
    }

File: src/nodes/synthesizer.ts
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { MongoClient } from "mongodb";
import { GlobalState } from "../state.js";
import { logTokenUsage } from "../utils/token_tracker.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function runSynthesizer(state: Partial<GlobalState>): Promise<Partial<GlobalState>> {
  const codedSections = state.coded_sections || [];
  const enhancedPrompt = state.enhanced_prompt || "";
  const userPrompt = state.user_prompt || "AI Generated UI Layout";

  if (codedSections.length === 0) {
    console.log("\n[Synthesizer] No coded sections to synthesize!");
    return {
      pipeline_status: "failed",
      failure_reason: "No coded sections found to synthesize",
    };
  }

  // Filter to only successfully coded sections
  const successfulSections = codedSections.filter((s) => s.status === "success");

  // Use theme tokens from state
  let rootCss = state.theme_css || "";
  if (!rootCss) {
    console.warn("[Synthesizer] ⚠ No theme_css in state. Using default tokens.");
    rootCss = `:root {
  --bg-primary: #0f172a;
  --bg-secondary: #1e293b;
  --bg-tertiary: #334155;
  --text-primary: #f8fafc;
  --text-secondary: #94a3b8;
  --accent-color: #3b82f6;
  --accent-hover: #2563eb;
  --font-family: 'Inter', sans-serif;
  --font-heading: 'Inter', sans-serif;
  --spacing-xs: 4px;
  --spacing-sm: 8px;
  --spacing-md: 16px;
  --spacing-lg: 32px;
  --spacing-xl: 64px;
  --border-radius: 8px;
  --border-radius-lg: 16px;
}`;
  }

  console.log("\n[Synthesizer] Using pre-extracted design tokens from theme_extractor.");

  const cssBlocks = successfulSections.map(
    (secCode) => `/* Section ${secCode.section_id}: ${secCode.section_name} */\n${secCode.css}`
  );

  // Extract Google Fonts to link/import
  const allCssToScan = rootCss + "\n" + cssBlocks.join("\n");
  const fontMatches = Array.from(
    allCssToScan.matchAll(/(?:font-family|font-heading|font-mono|--font-[a-z0-9-]+):\s*['"]?([^'"`\;,\)]+)['"]?/gi)
  ).map((m) => m[1].trim());

  let googleFontsImport = "";
  if (fontMatches.length > 0) {
    const fontFamilies: string[] = [];
    const uniqueFonts = Array.from(new Set(fontMatches));
    for (let font of uniqueFonts) {
      font = font.split(",")[0].trim().replace(/['"]/g, "");
      const systemFonts = ["sans-serif", "serif", "monospace", "cursive", "system-ui", "-apple-system", "inherit", "initial", "unset"];
      if (!systemFonts.includes(font.toLowerCase()) && !font.startsWith("var(")) {
        fontFamilies.push(font.replace(/\s+/g, "+"));
      }
    }
    if (fontFamilies.length > 0) {
      const fontQuery = fontFamilies.join("|");
      googleFontsImport = `@import url('https://fonts.googleapis.com/css2?family=${fontQuery}:wght@400;600;700;800&display=swap');\n\n`;
    }
  }

  console.log("[Synthesizer] Assembling React components and stylesheets...");

  const jsxComponents: string[] = [];
  const resolvedCms: Record<string, any> = {};
  const dbRecords: any[] = [];

  for (const secCode of successfulSections) {
    const sectionId = secCode.section_id;
    const sectionName = secCode.section_name;

    // Get raw JSX, strip internal React imports to prevent duplicate or invalid imports
    const rawJsx = secCode.jsx;
    let cleanJsx = rawJsx.replace(/^\s*import\s+.*?;?\s*$/gm, "");
    cleanJsx = cleanJsx.replace("export default function", "function");
    jsxComponents.push(`// --- Section ${sectionId}: ${sectionName} ---\n${cleanJsx}`);

    // CMS Resolution
    const sectionState = (state.sections || []).find((s) => s.id === sectionId);
    if (sectionState && sectionState.cms) {
      const cms = sectionState.cms;
      dbRecords.push(cms);

      const componentKeyCamel = sectionName
        .split(/[^a-zA-Z0-9]/)
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join("");
      const componentKey = componentKeyCamel.charAt(0).toLowerCase() + componentKeyCamel.slice(1);

      const elementsMap: Record<string, any> = {};
      for (const elem of cms.elements || []) {
        const elemName = elem.elementName;
        if (elemName) {
          elementsMap[elemName] = elem;
        }
      }
      resolvedCms[componentKey] = elementsMap;
    }
  }

  // Build combined JSX App Component
  const wrapperLines = [
    "export function GeneratedPage({ cmsData }) {",
    "  return (",
    "    <div className=\"generated-page-container\">",
  ];

  for (const secCode of successfulSections) {
    const sectionName = secCode.section_name;
    const componentName = sectionName
      .split(/[^a-zA-Z0-9]/)
      .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
      .join("");
    const propKey = componentName.charAt(0).toLowerCase() + componentName.slice(1);
    wrapperLines.push(`      <${componentName} cmsData={cmsData.${propKey}} />`);
  }

  wrapperLines.push("    </div>");
  wrapperLines.push("  );");
  wrapperLines.push("}");

  // Determine which React hooks are referenced across all generated JSX components
  const allJsxContent = jsxComponents.join("\n");
  const hooksToImport = ["useState", "useMemo", "useEffect"];
  const possibleHooks = [
    "useRef",
    "useCallback",
    "useContext",
    "useReducer",
    "useImperativeHandle",
    "useLayoutEffect",
    "useDebugValue",
  ];
  for (const hook of possibleHooks) {
    const hookRegExp = new RegExp(`\\b${hook}\\b`);
    if (hookRegExp.test(allJsxContent)) {
      hooksToImport.push(hook);
    }
  }

  const hooksImportList = Array.from(new Set(hooksToImport)).sort().join(", ");

  const projectIdVal = state.project_id || "";

  const appWrapperCode = `export default function App() {
  const [cmsData, setCmsData] = useState(cmsDataRaw.resolved_cms);
  const [editMode, setEditMode] = useState(false);
  const [hasChanges, setHasChanges] = useState(false);
  const [pendingChanges, setPendingChanges] = useState({});
  const [loading, setLoading] = useState(true);
  
  const projectId = "__PROJECT_ID__";

  useEffect(() => {
    fetch(\`http://localhost:5001/api/cms/\${projectId}\`)
      .then(res => {
        if (!res.ok) throw new Error("Server not running or project not found");
        return res.json();
      })
      .then(data => {
        if (data.resolved_cms) {
          setCmsData(data.resolved_cms);
        }
        setLoading(false);
      })
      .catch(err => {
        console.warn("Using local cms_data.json fallback:", err);
        setLoading(false);
      });
  }, []);

  useEffect(() => {
    const handleDblClick = (e) => {
      if (!editMode) return;
      const fieldId = e.target.getAttribute('data-field-id');
      if (fieldId) {
        e.target.contentEditable = 'true';
        e.target.focus();
      }
    };

    const handleBlur = (e) => {
      const fieldId = e.target.getAttribute('data-field-id');
      if (fieldId) {
        const newText = e.target.innerText.trim();
        setPendingChanges(prev => ({ ...prev, [fieldId]: newText }));
        setHasChanges(true);
      }
    };

    document.addEventListener('dblclick', handleDblClick);
    document.addEventListener('focusout', handleBlur);
    return () => {
      document.removeEventListener('dblclick', handleDblClick);
      document.removeEventListener('focusout', handleBlur);
    };
  }, [editMode]);

  const handleSave = async () => {
    try {
      const res = await fetch(\`http://localhost:5001/api/cms/\${projectId}\`);
      if (!res.ok) throw new Error("Failed to contact server for saving");
      const data = await res.json();
      const records = data.db_records;

      let updatedCount = 0;
      for (const record of records) {
        let sectionUpdated = false;
        
        for (const elem of record.elements) {
          if (pendingChanges[elem.fieldId] !== undefined) {
            elem.content = pendingChanges[elem.fieldId];
            sectionUpdated = true;
            updatedCount++;
          }
          if (elem.loop && Array.isArray(elem.loop)) {
            for (const item of elem.loop) {
              for (let i = 1; i <= 10; i++) {
                const fId = item[\`fieldId\${i}\`];
                if (fId && pendingChanges[fId] !== undefined) {
                  item[\`field\${i}\`] = pendingChanges[fId];
                  sectionUpdated = true;
                  updatedCount++;
                }
              }
            }
          }
        }

        if (sectionUpdated) {
          const sectionId = record.metadata.sectionId;
          const updateRes = await fetch(\`http://localhost:5001/api/cms/\${projectId}/section/\${sectionId}/update\`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ elements: record.elements })
          });
          if (!updateRes.ok) throw new Error(\`Failed to update section \${sectionId}\`);
        }
      }

      alert(\`Successfully saved \${updatedCount} changes to MongoDB!\`);
      setHasChanges(false);
      setPendingChanges({});
      
      const refreshRes = await fetch(\`http://localhost:5001/api/cms/\${projectId}\`);
      const refreshData = await refreshRes.json();
      setCmsData(refreshData.resolved_cms);
    } catch (err) {
      console.error(err);
      alert("Error saving changes: " + err.message);
    }
  };

  return (
    <div className={\`app-wrapper \${editMode ? 'edit-mode-active' : ''}\`}>
      <div style={{
        position: 'fixed',
        bottom: '24px',
        right: '24px',
        zIndex: 99999,
        background: '#1e293b',
        border: '1px solid #38bdf8',
        padding: '12px 18px',
        borderRadius: '12px',
        boxShadow: '0 10px 15px -3px rgba(0, 0, 0, 0.3)',
        display: 'flex',
        gap: '12px',
        alignItems: 'center',
        fontFamily: 'sans-serif'
      }}>
        <span style={{ color: '#f8fafc', fontSize: '14px', fontWeight: 'bold' }}>
          {editMode ? '✍️ Edit Mode Active (Double Click Text to Edit)' : '👁️ Preview Mode'}
        </span>
        <button 
          onClick={() => setEditMode(!editMode)}
          style={{
            background: editMode ? '#ef4444' : '#38bdf8',
            color: '#0f172a',
            border: 'none',
            padding: '6px 12px',
            borderRadius: '6px',
            cursor: 'pointer',
            fontWeight: 'bold',
            transition: 'background 0.2s'
          }}
        >
          {editMode ? 'Disable Edit' : 'Enable Edit'}
        </button>
        {hasChanges && (
          <button 
            onClick={handleSave}
            style={{
              background: '#22c55e',
              color: '#ffffff',
              border: 'none',
              padding: '6px 12px',
              borderRadius: '6px',
              cursor: 'pointer',
              fontWeight: 'bold',
              transition: 'background 0.2s'
            }}
          >
            Save to MongoDB
          </button>
        )}
      </div>

      <style>{\`
        .edit-mode-active [data-field-id] {
          outline: 1px dashed #38bdf8 !important;
          cursor: text !important;
          position: relative !important;
          z-index: 10000 !important;
        }
        .edit-mode-active [data-field-id]:hover {
          background: rgba(56, 189, 248, 0.1) !important;
        }
        .edit-mode-active [data-field-id]:focus {
          outline: 2px solid #38bdf8 !important;
          background: rgba(56, 189, 248, 0.15) !important;
        }
      \`}</style>

      <GeneratedPage cmsData={cmsData} />
    </div>
  );
}`.replace("__PROJECT_ID__", projectIdVal);

  const combinedJsx =
    `import React, { ${hooksImportList} } from 'react';\n` +
    `import cmsDataRaw from './cms_data.json';\n` +
    `import './index.css';\n\n` +
    jsxComponents.join("\n\n") + "\n\n" +
    wrapperLines.join("\n") + "\n\n" +
    appWrapperCode;

  const combinedCss = googleFontsImport + rootCss + "\n\n" + cssBlocks.join("\n\n");

  const cmsPayload = {
    db_records: dbRecords,
    resolved_cms: resolvedCms,
  };

  const projectRoot = path.join(__dirname, "..", "..");

  const outputDir = path.join(projectRoot, "output");
  fs.mkdirSync(outputDir, { recursive: true });
  const outJsxFile = path.join(outputDir, "page.jsx");
  const outCssFile = path.join(outputDir, "page.css");
  const outCmsFile = path.join(outputDir, "cms_data.json");

  const reactSrcDir = path.join(projectRoot, "testing_react", "src");
  fs.mkdirSync(reactSrcDir, { recursive: true });
  const reactJsxFile = path.join(reactSrcDir, "App.jsx");
  const reactCssFile = path.join(reactSrcDir, "index.css");
  const reactCmsFile = path.join(reactSrcDir, "cms_data.json");

  try {
    // Write to output folder
    fs.writeFileSync(outJsxFile, combinedJsx, "utf-8");
    fs.writeFileSync(outCssFile, combinedCss, "utf-8");
    fs.writeFileSync(outCmsFile, JSON.stringify(cmsPayload, null, 2), "utf-8");

    // Write to testing_react project
    fs.writeFileSync(reactJsxFile, combinedJsx, "utf-8");
    fs.writeFileSync(reactCssFile, combinedCss, "utf-8");
    fs.writeFileSync(reactCmsFile, JSON.stringify(cmsPayload, null, 2), "utf-8");

    console.log(`[Synthesizer] ✓ React components compiled to output: ${outJsxFile}`);
    console.log(`[Synthesizer] ✓ CSS stylesheet written to output: ${outCssFile}`);
    console.log(`[Synthesizer] ✓ Unified CMS JSON payload written to output: ${outCmsFile}`);
    console.log(`[Synthesizer] ✓ React components integrated into testing_react: ${reactJsxFile}`);
    console.log(`[Synthesizer] ✓ CSS stylesheet integrated into testing_react: ${reactCssFile}`);
    console.log(`[Synthesizer] ✓ CMS JSON integrated into testing_react: ${reactCmsFile}`);

    const mongodbUri = process.env.MONGODB_URI;
    if (mongodbUri) {
      try {
        const client = new MongoClient(mongodbUri);
        await client.connect();
        const db = client.db("UI-Layout-DB");
        const collection = db.collection("cms_data");

        const projectId = state.project_id || "";
        const projectName = state.user_prompt || "";

        for (const record of dbRecords) {
          const doc = {
            projectId: projectId,
            projectName: projectName,
            metadata: record.metadata || {},
            elements: record.elements || [],
          };
          await collection.insertOne(doc);
        }

        console.log(`[Synthesizer] ✓ Saved ${dbRecords.length} sections to MongoDB collection 'cms_data'`);
        await client.close();
      } catch (dbErr: any) {
        console.error(`[Synthesizer] ✗ Failed to save CMS records to MongoDB: ${dbErr.message}`);
      }
    }
  } catch (err: any) {
    console.error(`[Synthesizer] ✗ Failed to write output files: ${err.message}`);
    return {
      pipeline_status: "failed",
      failure_reason: `Writing synthesized files failed: ${err.message}`,
    };
  }

  return {
    final_html: combinedJsx,
    output_path: outJsxFile,
    pipeline_status: "complete",
  };
}
